### PR TITLE
Implement motion import

### LINKED ISSUE #1713: [imports] Relocate csv imports to the backend: Motion
Action description still needed.

### PR DESCRIPTION
**Pull request by luisa-beerboom**
Closes #1713 

Uses OpenSlides/openslides-meta#14

### CODE REVIEW

**Comment by luisa-beerboom** [comment]
Up until now the `json_upload` seems to work in all cases involving only the required fields (including reason, which is sometimes required), as well as the number field.

Submitter generation in case of omission seems to be working as well.

The categories have been tested in so far that the `json_upload` seems to work for most common cases (there are still some untested edge cases though)

There are some remaining TODOs and all other fields have not yet been tested at all.
The import too is neither finished, nor tested.

@r-peschke If you wish to have a look at and give feedback to the finished parts sometime next week, please feel free to do so, I'd appreciate that

File: openslides_backend/action/actions/motion/json_upload.py [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+            entry["motion_amendment"] = {
+                "value": is_amendment,
+                "info": ImportState.WARNING,
+            }
+            messages.append("Amendments cannot be correctly imported")
+
+        if (category_name := entry.get("category_name")) and isinstance(
+            category_name, str
+        ):
+            category_prefix = entry.get("category_prefix") or None
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:189]: You don't need   the `or None`. The get-method returns the value, if the dictionary has the attribute or it returns None.

File: openslides_backend/action/actions/motion/json_upload.py [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+            entry[fieldname] = {
+                "value": entry.get(fieldname, ""),
+                "info": ImportState.ERROR,
+            }
+        else:
+            entry[fieldname]["info"] = ImportState.ERROR
+        self.row_state = ImportState.ERROR
+        return entry
+
+    def _get_motion_mock_id(self) -> int:
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:716]: Please remove the method, an id is not needed

File: openslides_backend/action/actions/motion/json_upload.py [hunk 3]
```diff
@@ -0,0 +1,723 @@
...
+            data = self.datastore.filter(
+                collection,
+                And(
+                    *and_filters,
+                    Or([FilterOperator(fieldname, "=", name) for name in set(entries)]),
+                ),
+                [*mapped_fields, "id", fieldname],
+                lock_result=False,
+            )
+            for date_id in data:
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:640]: Replace the line by `for date in data.values():`

**Review by r-peschke — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/motion/json_upload.py (hunk 1 above)
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py:189]: I need `or None` because I need this to ignore empty strings

File: openslides_backend/action/actions/motion/json_upload.py [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+
+    def get_lookup_dict(
+        self,
+        collection: str,
+        entries: List[str],
+        fieldname: str = "name",
+        mapped_fields: List[str] = [],
+        and_filters: List[Filter] = [],
+    ) -> Dict[str, List[Dict[str, Any]]]:
+        lookup: Dict[str, List[Dict[str, Any]]] = {}
```
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py:629]: There is nothing above

File: openslides_backend/action/actions/motion/json_upload.py (hunk 1 above)
Comment (r-peschke) [reply openslides_backend/action/actions/motion/json_upload.py:189]: Then please write a test for this edge case:-).

File: openslides_backend/action/actions/motion/json_upload.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/motion/json_upload.py:629]: > There is nothing above

Sorry, I meant **below**

File: openslides_backend/action/actions/motion/json_upload.py [hunk 5]
```diff
@@ -0,0 +1,723 @@
...
+    _first_state_id: Optional[int] = None
+    _operator_username: Optional[str] = None
+
+    _last_motion_mock_id: Optional[int] = None
+    _user_ids_to_meeting_user: Dict[int, Any]
+
+    def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
+        # transform instance into a correct create/update payload
+        # try to find a pre-existing motion with the same number
+        # if there is one, validate for a motion.create, otherwise for a motion.update
```
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py:134]: Seems like I wrote that the wrong way around, I'll change it

File: openslides_backend/action/actions/motion/json_upload.py (hunk 1 above)
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py:189]: Already exists

File: openslides_backend/action/actions/motion/json_upload.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py:629]: At what line of what file is that supposed to be? I can't find it

**Review by r-peschke — CHANGES_REQUESTED** [review]
Some small things.

File: openslides_backend/action/actions/motion/json_upload.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: ?

File: openslides_backend/action/actions/motion/json_upload.py [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+                        elif user in username_set:
+                            entry_list.append(
+                                {"value": user, "info": ImportState.WARNING}
+                            )
+                            message_set.add(
+                                f"At least one {fieldname} has been referenced multiple times"
+                            )
+                        else:
+                            username_set.add(user)
+                            found_users = lookup.get(user, [])
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:290]: You cannot get more than one user, the username is unique. If you find duplicate users, you should throw an exception and finish the json_upload immediately,

File: openslides_backend/action/actions/motion/json_upload.py [hunk 7]
```diff
@@ -0,0 +1,708 @@
...
+                    if isinstance(user, str):
+                        if verbose_user_mismatch:
+                            entry_list.append(
+                                {"value": user, "info": ImportState.ERROR}
+                            )
+                        elif user in username_set:
+                            entry_list.append(
+                                {"value": user, "info": ImportState.WARNING}
+                            )
+                            message_set.add(
```
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py:285]: Done for users

File: openslides_backend/action/actions/motion/json_upload.py (hunk 6 above)
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py:290]: Done

File: openslides_backend/action/actions/motion/json_upload.py [hunk 8]
```diff
@@ -0,0 +1,709 @@
...
+    submitter_lookup: Dict[str, List[Dict[str, Any]]] = {}
+    supporter_lookup: Dict[str, List[Dict[str, Any]]] = {}
+    category_lookup: Dict[str, List[Dict[str, Any]]] = {}
+    tags_lookup: Dict[str, List[Dict[str, Any]]] = {}
+    block_lookup: Dict[str, List[Dict[str, Any]]] = {}
+    _first_state_id: Optional[int] = None
+    _operator_username: Optional[str] = None
+    _previous_numbers: List[str]
+
+    _last_motion_mock_id: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: Done

File: openslides_backend/action/actions/motion/json_upload.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: Done for tags

File: openslides_backend/action/actions/motion/json_upload.py (hunk 1 above)
Comment (r-peschke) [reply openslides_backend/action/actions/motion/json_upload.py]: If the `and None` is necessary, there should be a reason why it is necessary. I would expect removing it, changes the behaviour. And this change should be manifested in a failing test

Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: I removed the or None and it still works, seems it was unnecessary

File: openslides_backend/action/actions/motion/json_upload.py [hunk 9]
```diff
@@ -0,0 +1,712 @@
...
+                                *entry.get("supporters_username", []),
+                            ]
+                        )
+                        for username in usernames
+                        if username
+                    ]
+                )
+            ),
+            "username",
+            ["meeting_user_ids"],
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:573]: add the field `meeting_ids` to the mapped_fields. This field is the calculated field for membership in a meeting. You don't need to look for the group-membership and consequently this lookup is everything you need concerning the submitter- and supporter-users.

File: openslides_backend/action/actions/motion/json_upload.py [hunk 10]
```diff
@@ -0,0 +1,712 @@
...
+        )
+        all_user_ids = [
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:575]: All this code beginning here down to line 609 is not needed, see remark above

File: openslides_backend/action/actions/motion/json_upload.py [hunk 11]
```diff
@@ -0,0 +1,708 @@
...
+                other_forbidden_numbers=self._previous_numbers,
+            )
+            if number := value.get("number"):
+                entry["number"] = {"value": number, "info": ImportState.GENERATED}
+                self._previous_numbers.append(number)
+
+        has_submitter_error: bool = False
+        for fieldname in ["submitter", "supporter"]:
+            if users := entry.get(f"{fieldname}s_username"):
+                if not isinstance(users, list):
```
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: I already have tests that check if it works when the submitters/supporters are not in a list.
It works, even after removing the if clauses. I surmise that the if clauses were unnecessary.
Does this address your concerns for this remark?

File: openslides_backend/action/actions/motion/json_upload.py [hunk 12]
```diff
@@ -0,0 +1,712 @@
...
+                message_map: Dict[str, Tuple[str, List[str]]] = {}
+                for user in users:
+                    if verbose_user_mismatch:
+                        entry_list.append({"value": user, "info": ImportState.ERROR})
+                    elif user in username_set:
+                        entry_list.append({"value": user, "info": ImportState.WARNING})
+                        err_message, err_users = message_map.get(
+                            "duplicate",
+                            (
+                                f"At least one {fieldname} has been referenced multiple times: ",
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:269]: It's better to gather the `username` in a list,  together with the `fieldname` and create the message one time at the end of the `fieldnames` loop

**Review by r-peschke — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/motion/json_upload.py (hunk 9 above)
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: Done in json_upload

**Comment by luisa-beerboom** [comment]
TODO:

see  #2062

File: openslides_backend/action/actions/motion/json_upload.py [hunk 13]
```diff
@@ -0,0 +1,708 @@
...
+                    users = [users]
+                verbose = entry.get(f"{fieldname}s_verbose", [])
+                if not isinstance(verbose, list):
+                    verbose = [verbose]
+                verbose_user_mismatch = len(verbose) > len(users)
+                username_set: Set[str] = set([])
+                entry_list: list[Dict[str, Any]] = []
+                message_set = set()
+                lookup = (
+                    self.submitter_lookup
```
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: Done in import as well

File: openslides_backend/action/actions/motion/json_upload.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: Done in import as well

File: openslides_backend/action/actions/motion/json_upload.py (hunk 9 above)
Comment (luisa-beerboom) [reply openslides_backend/action/actions/motion/json_upload.py]: Done in import as well

**Comment by luisa-beerboom** [comment]
Rewrote json_upload tests, found some things about the main class that I thought needed changing/reconsidering, so there are currently some TODOs in the tests and some of them fail (as I've already rewritten them).

Also I am planning to move many of these tests to a separate class for the purpose of using their results in the import test that I've yet to rewrite, so no need to worry about that yet.

@r-peschke would you take a look at what I have so far?

File: openslides_backend/action/actions/motion/json_upload.py [hunk 14]
```diff
@@ -0,0 +1,607 @@
...
+                mapped_fields=["first_state_id"],
+            ).values()
+            if len(default_workflows) != 1:
+                raise ActionException("Couldn't determine default workflow")
+            self._first_state_id = cast(
+                int, list(default_workflows)[0].get("first_state_id")
+            )
+        return self._first_state_id
+
+    def _get_field_array(self, entry: dict[str, Any], fieldname: str) -> list[str]:
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:567]: Seems, that this method is never used

File: openslides_backend/action/actions/motion/json_upload.py [hunk 15]
```diff
@@ -0,0 +1,607 @@
...
+    def _get_field_id(self, entry: dict[str, Any], fieldname: str) -> int:
+        return entry[fieldname].get("id")
+
+    def _add_error_to_entry(
+        self, entry: dict[str, Any], err: MotionActionErrorData
+    ) -> dict[str, Any]:
+        fieldname = ""
+        match err["type"]:
+            case MotionErrorType.UNIQUE_NUMBER:
+                fieldname = "number"
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/json_upload.py:590]: Could you write a test for this case?

**Review by r-peschke — CHANGES_REQUESTED** [review]
Sorry, but there are some things.
Tomorrow I'll continue with import, ...

File: openslides_backend/action/actions/motion/import_.py [hunk 16]
```diff
@@ -0,0 +1,644 @@
...
+                        )
+                        old_submitters[id_] = matched_submitters
+
+                    payload.pop("meeting_id", None)
+                    update_action_payload.append(payload)
+                    errors = self.get_update_payload_integrity_error_message(
+                        payload, meeting_id
+                    )
+                for err in errors:
+                    fieldname = ""
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/import_.py:196]: Maybe the problem will be resolved by other tests, but  there is no test with errors passing this code

File: openslides_backend/action/actions/motion/import_.py [hunk 17]
```diff
@@ -0,0 +1,644 @@
...
+                        f"Invalid JsonUpload data: A data row with state '{ImportState.DONE}' must have an 'id'"
+                    )
+                elif entry["id"] != id_:
+                    row["state"] = ImportState.ERROR
+                    entry["number"]["info"] = ImportState.ERROR
+                    row["messages"].append(
+                        f"Error: Number '{number}' found in different id ({id_} instead of {entry['id']})"
+                    )
+            elif check_result == ResultType.FOUND_MORE_IDS:
+                row["state"] = ImportState.ERROR
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/import_.py:326]: write tests to provoke this error. There are different combinations possible, i.e 2 on database (technically possible), 2 in payload

File: openslides_backend/action/actions/motion/import_.py [hunk 18]
```diff
@@ -0,0 +1,644 @@
...
+
+        number = self.get_value_from_union_str_object(entry.get("number"))
+        if number:
+            check_result = self.number_lookup.check_duplicate(number)
+            id_ = cast(int, self.number_lookup.get_field_by_name(number, "id"))
+
+            if check_result == ResultType.FOUND_ID and id_ != 0:
+                if row["state"] != ImportState.DONE:
+                    row["messages"].append(
+                        f"Error: Row state expected to be '{ImportState.DONE}', but it is '{row['state']}'."
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/import_.py:311]: Missing test for this error

File: openslides_backend/action/actions/motion/import_.py [hunk 19]
```diff
@@ -0,0 +1,644 @@
...
+                    )
+                    row["state"] = ImportState.ERROR
+                    entry["number"]["info"] = ImportState.ERROR
+                elif "id" not in entry:
+                    raise ActionException(
+                        f"Invalid JsonUpload data: A data row with state '{ImportState.DONE}' must have an 'id'"
+                    )
+                elif entry["id"] != id_:
+                    row["state"] = ImportState.ERROR
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/import_.py:320]: Missing test for this error

File: openslides_backend/action/actions/motion/import_.py [hunk 20]
```diff
@@ -0,0 +1,644 @@
...
+                            tag_entry["info"] = ImportState.ERROR
+                            tag_entry.pop("id")
+                            different.append(tag)
+                    else:
+                        tag_entry["info"] = ImportState.ERROR
+                        tag_entry.pop("id")
+                        not_found.append(tag)
+            if len(different):
+                row["messages"].append(
+                    "Error: Tag search didn't deliver the same result as in the preview: "
```
Comment (r-peschke) [inline openslides_backend/action/actions/motion/import_.py:421]: missing test for this error

**Review by r-peschke — CHANGES_REQUESTED** [review]

File: tests/system/action/motion/test_create.py [hunk 21]
```diff
@@ -148,6 +148,43 @@ def test_create_workflow_id(self) -> None:
...
+        response = self.request(
+            "motion.create",
+            {
+                "title": "title_test1",
+                "meeting_id": 222,
+                "text": "test",
+            },
+        )
+        self.assert_status_code(response, 200)
+        motion = self.get_model("motion/1")
```
Comment (r-peschke) [inline tests/system/action/motion/test_create.py:182]: Here you should use the self.assert_model_exists-method

File: openslides_backend/action/actions/motion/update.py [hunk 22]
```diff
@@ -93,34 +98,17 @@ def prefetch(self, action_data: ActionData) -> None:
...
-                    "Cannot update text, because it was not set in the old values."
-                )
+        motion = self.datastore.get(
+            fqid_from_collection_and_id(self.model.collection, instance["id"]),
+            ["meeting_id"],
+        )
+        error_messages = self.get_update_payload_integrity_error_message(
+            instance, motion["meeting_id"]
+        )
+        if len(error_messages):
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/motion/update.py:112]: Then there is no need to continue after first error

[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/motion/update.py:112]: closed

**Review by r-peschke — APPROVED** [review]
